flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/PipelinedSubpartition.java

```diff
@@ -315,6 +326,17 @@ BufferAndBacklog pollBuffer() {
                 if (buffer.readableBytes() > 0) {
                     break;
                 }
+
+                // if we have an empty finished buffer and the exclusive credit is 0, we just return
+                // the empty buffer so that the downstream task can release the allocated credit for
+                // this empty buffer, this happens in two main scenarios currently:
+                // 1. all data of a buffer builder has been read and after that the buffer builder
+                // is finished
+                // 2. in approximate recovery mode, a partial record takes a whole buffer builder
+                if (buffersPerChannel == 0 && bufferConsumer.isFinished()) {
+                    break;
+                }
+
```

Comment:
Instead of this check here, can not we add similar check outside of the `while (...)` loop? For example replace:

```
if (buffer == null) {
    return null;
}
```
with something like:
```
if (buffer == null) {
   if (buffersPerChannel == 0) {
       return EMPTY_BUFFER;
    }
    else {
       return null;
    }
}
```
?
That way, we would avoid sending empty buffer if there are still more buffers in the backlog that are already enqueued?

Reply:
I am afraid we can not skip the empty buffer is the exclusive credit is 0, because we may already allocate the floating credit for it and we can not just release the floating credit because we may already announce the credit to the upstream.

Reply:
I was not suggesting that. I was suggesting to:
> we would avoid sending empty buffer if there are still more buffers in the backlog that are already enqueued

Reply:
I guess I did not fully understand the comment, could you please explain a bit more? From my understand, even there are multiple buffers in the queue and the first is empty, if we do not send the first empty buffer out, how could we guarantee the downstream tasks are not allocating too many buffers if the exclusive credit is 0? For example, if there is two buffers, the first one is empty and the second one is an event, we may already announce 1 backlog to the downstream task and if we send the event directly, dose that mean we can not release the credit already allocated? Correct me if I am wrong.

Reply:
If there are two buffers, first is empty the second one can be:
1. non empty buffer
2. empty buffer
3. event

In all cases, instead of sending this first empty buffer with the backlog information, as far as I understand, it should be possible to just send any of the above instead with the updated backlog information. From the downstream node perspective, it should make no difference if we hide this empty buffer. If this is 2nd or 3rd case, we should be able to release the floating buffer regardless of that, shouldn't we?

Reply:
Let's maybe focus on the 3rd case first and we assume that the exclusive credit is 0.

1. There are only one data buffer in the queue.
2. Flush triggered.
3. All data of the first buffer is committed but the buffer is still not finished.
4. All data of the buffer is consumed by pollBuffer and the available credit becomes 0.
5. The first buffer is finished, the second event is added and the data available notification is triggered.
6. The upstream announces backlog to the downstream to request a credit.
7.  The upstream receives available credit and start to pollBuffer.
8. Skip the first empty buffer and send the second event.
9. The downstream receive the event but the event does not consume any credit.

Do you mean we should change the current logic and release the floating buffer for event in some cases (including reduce the available credit by 1 at the upstream, currently the available credit is not decreased for event)? If there are multiple empty buffers, should we just skip the first one or should we skip all?